File: backend/src/main/java/com/weddingraffle/rifa/service/TransactionService.java
package com.weddingraffle.rifa.service;

import com.weddingraffle.rifa.dto.TransactionCreateRequest;
import com.weddingraffle.rifa.dto.TransactionCreateResponse;
import com.weddingraffle.rifa.dto.TransactionQuoteRequest;
import com.weddingraffle.rifa.dto.TransactionQuoteResponse;
import com.weddingraffle.rifa.dto.TransactionRecoveryRequest;
import com.weddingraffle.rifa.dto.TransactionStatusResponse;

public interface TransactionService {

    TransactionQuoteResponse quote(TransactionQuoteRequest request);

    TransactionCreateResponse create(String idempotencyKey, TransactionCreateRequest request);

    void processPaymentNotification(String paymentId);

    TransactionStatusResponse getStatus(String externalReference, String paymentId);

    TransactionStatusResponse recover(TransactionRecoveryRequest request);
}

File: frontend/src/services/transactionService.ts
import { apiClient } from '../config/apiClient';
import type {
  TransactionCreateRequest,
  TransactionCreateResponse,
  TransactionQuoteRequest,
  TransactionQuoteResponse,
  TransactionRecoveryRequest,
  TransactionStatusResponse,
} from '../types/transaction';

export const transactionService = {
  async quote(request: TransactionQuoteRequest): Promise<TransactionQuoteResponse> {
    const response = await apiClient.post<TransactionQuoteResponse>('/transactions/quote', request);
    return response.data;
  },

  async create(
    request: TransactionCreateRequest,
    idempotencyKey: string,
  ): Promise<TransactionCreateResponse> {
    const response = await apiClient.post<TransactionCreateResponse>('/transactions', request, {
      headers: { 'Idempotency-Key': idempotencyKey },
    });
    return response.data;
  },

  async getStatus(
    externalReference: string,
    signal?: AbortSignal,
  ): Promise<TransactionStatusResponse> {
    const response = await apiClient.get<TransactionStatusResponse>(
      `/transactions/${externalReference}/status`,
      { signal },
    );
    return response.data;
  },

  async recover(request: TransactionRecoveryRequest): Promise<TransactionStatusResponse> {
    const response = await apiClient.post<TransactionStatusResponse>(
      '/transactions/recovery',
      request,
    );
    return response.data;
  },

  getLuckyNumbersPdfUrl(externalReference: string): string {
    return `${apiClient.defaults.baseURL}/transactions/${externalReference}/lucky-numbers.pdf`;
  },
};
